File: java/src/main/java/com/searchio/interview/fetcher/RemoteStoreFetcher.java
package com.searchio.interview.fetcher;

import com.searchio.interview.model.Entity;

import java.util.Map;

/**
 * Yes, this implementation is just backed by a {@link Map}, but really it might be back by a remote store
 * an RDBMS, REST api ... something that takes time.
 * A simple implementation of a Fetcher backed by a Map. This
 */
public class RemoteStoreFetcher implements Fetcher {

    private final Map<Long, Entity> entityMap;

    public RemoteStoreFetcher(Map<Long, Entity> entityMap) {
        this.entityMap = entityMap;
    }

    @Override
    public Entity get(long id) {
        try {
            Thread.sleep(4000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        return entityMap.get(id);
    }

    @Override
    public Entity save(Entity entity) {
        try {
            Thread.sleep(4000);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
        entityMap.put(entity.getIdentifier(), entity);
        return entity;
    }
}

File: csharp/SearchIOCodingExercise/CachingExercise/Fetcher/RemoteStoreFetcher.cs
using System.Collections.Generic;
using System.Threading;
using CachingExercise.Model;

namespace CachingExercise.Fetcher
{
    ///<summary>
    /// Yes, this implementation is just backed by a Dictionary, but really it might be backed by a remote store
    /// an RDBMS, REST api ... something that takes time. There is an arbitrary wait in this implementation to
    /// signify that this class is slow.
    /// </summary>
    public class RemoteStoreFetcher : IFetcher
    {
        private Dictionary<long, IEntity> _entityStore;

        public RemoteStoreFetcher(Dictionary<long, IEntity> entityStore)
        {
            _entityStore = entityStore;
        }

        public IEntity Get(long id)
        {
            Thread.Sleep(4000);
            return _entityStore.TryGetValue(id, out var value) ? value : null;
        }

        public IEntity Save(IEntity entity)
        {
            Thread.Sleep(4000);
            _entityStore.Add(entity.GetIdentifier(), entity);
            return entity;
        }

        public void Clear()
        {
            _entityStore.Clear();
        }
    }
}